cluster-autoscaler/utils/kubernetes/listers.go

```diff
@@ -46,6 +57,19 @@ type ListerRegistry interface {
 	JobLister() v1batchlister.JobLister
 	ReplicaSetLister() v1appslister.ReplicaSetLister
 	StatefulSetLister() v1appslister.StatefulSetLister
+	GenericListerFactory() GenericListerFactory
```

Comment:
What do we need such two-level interface for? Why not just have `GenericLister(gvr schema.GroupVersionResource, namespace string) dynamiclister.Lister` directly here?

Reply:
The main reason is so that tests can implement this interface (like [this](https://github.com/kubernetes/autoscaler/pull/5419/files#diff-35dba75ac3715863c6460ab9d5515b582d1d05d705b5e597bb25a511ed42171fR177-R195) for example). I am still going through the unit tests. This interface might not be necessary but when I look at the tests it seems like having this would be useful. I would be happy to remove it if I am wrong because like you said not having the interface simplifies the code.

Reply:
Also, having a factory like this means, `GenericLister` can be used independently of the `ListerRegistry` (not sure if we have a use-case like that in cluster-autoscaler but I thought having that de-coupling might prove useful in the future.